nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -291,30 +347,197 @@ impl DataStore {
+                    // policy = 'fail'.
+                    // - AA,P=Allow: Same as above, but with policy = 'allow'.
+                    // - A,P=Fail: All sleds sharing an affinity instance within a group with
+                    // policy = 'fail'.
+                    // - A,P=Allow: Same as above, but with policy = 'allow'.
+                    //
+                    // ## Affinity: Apply Rules
+                    //
+                    // - Targets := All viable sleds for instance placement
+                    // - Banned := AA,P=Fail
+                    // - Required := A,P=Fail
+                    // - if Required.len() > 1: Fail (too many constraints).
+                    // - if Required.len() == 1...
+                    //   - ... if the entry exists in the "Banned" set: Fail
+                    //     (contradicting constraints 'Banned' + 'Required')
+                    //   - ... if the entry does not exist in "Targets": Fail
+                    //     ('Required' constraint not satisfiable)
+                    //   - ... if the entry does not exist in "Banned": Use it.
+                    //
+                    // If we have not yet picked a target, we can filter the
+                    // set of targets to ignore "banned" sleds, and then apply
+                    // preferences.
+                    //
+                    // - Targets := Targets \ Banned
+                    //
+                    // ## Affinity: Apply Preferences
+                    //
+                    // - Preferred := Targets ∩ A,P=Allow
+                    // - Unpreferred := Targets ∩ AA,P=Allow
+                    // - Neither := Preferred ∩ Unpreferred
+                    // - Preferred := Preferred \ Neither
+                    // - Unpreferred := Unpreferred \ Neither
+                    // - If Preferred isn't empty, pick a target from it.
+                    // - Targets := Targets \ Unpreferred
+                    // - If Targets isn't empty, pick a target from it.
+                    // - If Unpreferred isn't empty, pick a target from it.
+                    // - Fail, no targets are available.
+
+                    let mut targets: HashSet<_> = sled_targets.into_iter().collect();
+
+                    let banned = anti_affinity_sleds.iter().filter_map(|(policy, id)| {
+                        if *policy == AffinityPolicy::Fail {
+                            Some(*id)
+                        } else {
+                            None
+                        }
+                    }).collect::<HashSet<_>>();
+                    let required = affinity_sleds.iter().filter_map(|(policy, id)| {
+                        if *policy == AffinityPolicy::Fail {
+                            Some(*id)
+                        } else {
+                            None
+                        }
+                    }).collect::<HashSet<_>>();
```

Comment:
silly code golf nit, feel free to ignore: if you wanted to only iterate over `anti_affinity_sleds`/`affinity_sleds` a single time to get both the rules and preferences, you could maybe consider using [`Iterator::partition`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) to do it in one pass. But, that's a bit annoying because you can't throw away the policy value in the `partition` closure and there's no `partition_map` type operation yet...

Reply:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map exists... I can do this if we're cool with the itertools dependency?

Reply:
(Updated to use `partition_map`)